tests/partitions_test.py

```diff
@@ -0,0 +1,104 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+
+import infra.network
+import infra.e2e_args
+import infra.partitions
+import infra.logging_app as app
+import suite.test_requirements as reqs
+
+from loguru import logger as LOG
+
+
+@reqs.description("Invalid partitions are not allowed")
+def test_invalid_partitions(network, args):
+    nodes = network.get_joined_nodes()
+
+    try:
+        network.partitioner.partition(
+            [nodes[0], nodes[2]],
+            [nodes[1], nodes[2]],
+        )
+        assert False, "Node should not appear in two or more partitions"
+    except ValueError:
+        pass
+
+    try:
+        network.partitioner.partition()
+        assert False, "At least one partition should be specified"
+    except ValueError:
+        pass
+
+    try:
+        invalid_local_node_id = -1
+        new_node = infra.node.Node(invalid_local_node_id, "local://localhost")
+        network.partitioner.partition([new_node])
+        assert False, "All nodes should belong to network"
+    except ValueError:
+        pass
+
+    return network
+
+
+@reqs.description("Partition primary + f nodes")
+def test_partition_majority(network, args):
+    primary, backups = network.find_nodes()
+
+    # Create a partition with primary + half remaining nodes (i.e. majority)
+    partition = [backups[0]]
+    partition.extend(backups[len(backups) // 2 :])
+
+    # Note: Context manager
+    with network.partitioner.partition(partition):
+        try:
+            network.wait_for_new_primary(primary.node_id)
+            assert False, "No new primary should be elected when partitioning majority"
+        except TimeoutError:
+            pass
+
+    return network
+
+
+@reqs.description("Isolate primary")
+def test_isolate_primary(network, args):
+    primary, backups = network.find_nodes()
+
+    # Isolate first backup from primary so that first backup becomes candidate
+    # in a new term and wins the election
+    # Note: Managed manually
+    rules = network.partitioner.isolate_node(primary, backups[0])
+
+    network.wait_for_new_primary(backups[0].node_id)
+
+    # Explicitly drop rules before continuing
+    rules.drop()
```

Comment:
Again I may be missing something about the partitioning logic here, but I think the comment's inaccurate. If we partition the primary and the first backup, then if there's a majority of _remaining_ backups, one of them will win an election. That must be what's happening - `wait_for_new_primary(backups[0].node_id)` means wait for anyone _except `backups[0]`_ to become primary. I'm also not sure why this is managed manually. But we don't gain anything by partitioning `backups[0]` as well, that just increases the required number of nodes (equivalent to crashing 2 nodes rather only 1 primary)?

```suggestion
    # Isolate first backup from primary so that first backup becomes candidate
    # in a new term and wins the election
    # Note: Managed manually
    rules = network.partitioner.isolate_node(primary)

    network.wait_for_new_primary(primary.node_id)

    # Explicitly drop rules before continuing
    rules.drop()
```

Reply:
(Ignore this, I'd completely missed that this call was `isolate_node` rather than `partition`)